Add EKF support for multiple GPS receivers

priseborough opened the pull request:
**Problem**

The existing logic used to switch between GPS receivers can cause large unpredictable jumps in EKF position if dual GPs receivers are fitted and GPS_AUTO_SWITCH is set to 1.

**Solution**

Extend the GPS_AUTO_SWITCH to type 2 which is a soft switching/blending method using a weighted average of physical receivers to create a third 'virtual' GPS where:

1) The method is scaleable to more than 2 GPS receivers
2) A 'blended' location and velocity is calculated using a weighted average of the two receivers
3) Weights are calculated using the ratio of reported error variances.
4) A filtered location  offset correction is calculated for each receiver relative to the 'blended' location and used to correct the output from each receiver.
5) If less than 2 GPS units are available the first GPS instance detected will be used.
6) If there is not enough error reporting to support calculation of the blending weights, the first receiver with the equal highest solution status value will be used . This can happen if one of the two receivers does not provide a position or speed accuracy.

The existing hard switch logic has been left untouched and is selected with GPS_AUTO_SWITCH = 1 the same as before.

Logging has been updated to to log the blended GPS instance if it is being calculated.

**Status**

This is a draft for comment. it has had limited flight and bench testing.

**TODO**

Improve the legacy hard switch logic to use solution status and reported accuracy (if available) rather than number of satellites. Using the number of satellites doesn't work as a metric where there are mixed receiver models (eg GPS+GLONASS and GPS only).
Bench testing of various GPS failure modes - signal fade, loss of physical connection.

proficnc commented:
Thanks Paul
Looking forward to testing this!

priseborough commented:
I ran some bench tests here GPS1 (a UBlox 6) and GPS2 (a UBlox8) were alternately wrapped in foil to cause signal fade. The following plots show how GPS3 which is the 'virtual' blended GPS is largely unaffected.

Number of satellites reported - GPS 3 reports the largest value from any physical receivers being used in its solution:
![nsats](https://cloud.githubusercontent.com/assets/3596952/22414948/d5afc762-e717-11e6-9338-cc03fe60c3df.png)

Vertical Velocity
![vz](https://cloud.githubusercontent.com/assets/3596952/22414953/dbc70ec6-e717-11e6-9f64-0bd5d0af967e.png)

Altitude
![alt](https://cloud.githubusercontent.com/assets/3596952/22414956/e10867c2-e717-11e6-9812-55f3baad127a.png)

Latitude
![lat](https://cloud.githubusercontent.com/assets/3596952/22414966/e9f57cc6-e717-11e6-9bbb-d673b0117aa4.png)

Longitude
![lng](https://cloud.githubusercontent.com/assets/3596952/22414980/fb10ab8e-e717-11e6-94ac-72c7d1ff9695.png)

priseborough commented:
I'm currently investigating why disconnecting GPS2 eventually causes GPS1 data to stop updating.

magicrub commented:
Related to and will possibly close issue https://github.com/ArduPilot/ardupilot/issues/3547

Also related to and will possibly close issue https://github.com/ArduPilot/ardupilot/issues/3754

rmackay9 commented:
I'd definitely like to get this into AC3.5

rmackay9 commented:
@priseborough, this has a WIP label on it but maybe that's out of date?  Are you happy for this to go into master now?

OXINARF commented:
Was this reviewed by anyone? @WickedShell?

@rmackay9 By the way I don't think we should be including new things in the middle of beta testing 3.5. For 3.5.1 it would be fine, but putting new things now just delays more the final 3.5 release.

priseborough commented:
@rmackay9 I've been flying it regularly so am happy for it to go in if it has been reviewed so it can get some testing with different hardware.

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -213,6 +213,23 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
+    // @Param: BLEND_MASK
+    // @DisplayName: Multi GPS Blending Mask
+    // @Description: Determines which of the accuracy measures Horizontal position, Vertical Position and Speed are used to calculate the weighting on each GPS receiver when soft switching has been selected by setting GPS_AUTO_SWITCH to 3.
+    // @Values: 5:Default
```
Values and bitmask shouldn't be used in the same parameter documentation, in this case Values should be removed.

priseborough replied:
will fix

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -527,47 +557,70 @@ AP_GPS::update(void)
+                }
+                if (state[i].status > state[primary_instance].status) {
+                    // we have a higher status lock, change GPS
+                    primary_instance = i;
```
This is not really a new fault, but the last_instance_swap_ms should be updated here.

priseborough replied:
will fix

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +952,435 @@ float AP_GPS::get_lag(uint8_t instance) const
+            min_ms = state[i].last_gps_time_ms;
+        }
+        if (_rate_ms[i] > max_rate_ms) {
+            max_rate_ms = _rate_ms[i];
```
Does it matter that the _rate_ms is only the expected update rate? I primarily ask as the only people who will have this correct will be u-blox users, no other driver uses it, which means that every other driver is just configured to the rate it is hard coded to request or it receives (IE you can get 10Hz out of SBF, but the parameter won't reflect that unless a user realizes they should make it match.).

priseborough replied:
It does matter if the GPS is slower than the parameter - eg the parameter was set to 100ms and the GPs was updating at 200. It would likely result in the data from that GPS being ignored by the blending algorithm due to failing the data delayed check at line 1001.

Given that the default parameter value is 200msec, I don't think this is a problem as we would not want to use a GPS receiver that outputted slower than that anyway. If a user sets the parameter to a different value, then they are responsible for checking that their GPS is capable of that rate.

WickedShell replied:
@priseborough So my concern on this actually is that there are drivers right now where the driver requests 10Hz updates regardless of the parameter, and no user knows that they should be setting this parameter. I guess we could have some of those drivers instead call the set parameter to override to 10Hz which might work, but this is a very weak to incorrect assumption for any driver other then u-blox. (Indeed I was actually planning to hide this parameter inside the ublox layer with the parameter hiding stuff @tridge was talking about).

Would instead tracking the update time between instances in the AP_GPS layer as an actual calculated number be fine for your use case? One of the last things with #2879 is to actually track this in the AP_GPS layer so we can publish if it's healthy or not. We could track this as a delta update time that you could look into. If that's fine I can get that sent in relatively soon.

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +952,435 @@ float AP_GPS::get_lag(uint8_t instance) const
+    if (_blend_mask & USE_SPD_ACC) {
+        for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+            if (state[i].status >= GPS_OK_FIX_3D) {
+                if (state[i].have_speed_accuracy && state[i].speed_accuracy > 0.0f) {
```
Why check for speed_accuracy > 0.0f? Isn't that already assured via the have_speed_accuracy flag? It's a safe check, but I just don't see why you need it.

priseborough replied:
will fix

WickedShell commented on libraries/DataFlash/LogStructure.h:
```diff
@@ -783,6 +783,8 @@ Format characters in the format string for binary log messages
       "GPS2", "QBIHBcLLefffB", "TimeUS,Status,GMS,GWk,NSats,HDop,Lat,Lng,Alt,Spd,GCrs,VZ,U" }, \
+    { LOG_GPS3_MSG, sizeof(log_GPS), \
+       "GPS3", "QBIHBcLLefffB", "TimeUS,Status,GMS,GWk,NSats,HDop,Lat,Lng,Alt,Spd,GCrs,VZ,U" }, \
     { LOG_GPA_MSG,  sizeof(log_GPA), \
```
You need a GPA3, or you will be writing the blended GPS accuracies to IMU log entries.

priseborough replied:
will fix

WickedShell commented on libraries/AP_GPS/AP_GPS.h:
```diff
@@ -349,12 +361,14 @@ class AP_GPS
     AP_Int8 _raw_data;
-    AP_Int8 _gnss_mode[2];
-    AP_Int16 _rate_ms[2];
+    AP_Int8 _gnss_mode[GPS_MAX_INSTANCES];
```
This is a user facing control parameter, I don't see any value to the blending to be duplicating this.

priseborough replied:
will fix

priseborough replied:
This doesn't add a parameter because GPS_MAX_INSTANCES = 2

WickedShell commented on libraries/AP_GPS/AP_GPS.h:
```diff
@@ -349,12 +361,14 @@ class AP_GPS
-    AP_Int8 _gnss_mode[2];
-    AP_Int16 _rate_ms[2];
+    AP_Int8 _gnss_mode[GPS_MAX_INSTANCES];
+    AP_Int16 _rate_ms[GPS_MAX_INSTANCES];
```
Same deal, why have an extra parameter that doesn't appear to be utilized.

priseborough replied:
This doesn't add a parameter because GPS_MAX_INSTANCES = 2

WickedShell commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -710,7 +710,12 @@ void DataFlash_Class::Log_Write_GPS(const AP_GPS &gps, uint8_t i, uint64_t time_
-    const struct Location &loc = gps.location(i);
+    struct Location loc;
+    if (i >= gps.num_sensors()) {
+        loc = gps.location();
```
This seems really messy. Why not have gps.location(3) manage the blended instance directly within the GPS library.

priseborough replied:
will fix

priseborough commented:
Here are the plots for the physical (GPS and GPS2) and the virtual blended GPS3 receiver from todays testing of the updates. The accuracy reported for the blended GPS is the better of the two receivers as expected. The reported position and altitude of the blended receiver more closely follows the better receiver as expected

speed accuracy
![screen shot 2017-03-04 at 9 55 28 am](https://cloud.githubusercontent.com/assets/3596952/23572063/b7a69f12-00c0-11e7-9b32-1431bc8680b9.png)

position accuracy
![screen shot 2017-03-04 at 9 57 08 am](https://cloud.githubusercontent.com/assets/3596952/23572102/f270c5a0-00c0-11e7-809c-210af28f8e74.png)

Altitude
![screen shot 2017-03-04 at 9 57 29 am](https://cloud.githubusercontent.com/assets/3596952/23572111/fe766c6a-00c0-11e7-90d3-77e71ccf5fb9.png)

Latitude
![screen shot 2017-03-04 at 9 57 45 am](https://cloud.githubusercontent.com/assets/3596952/23572129/0be628b8-00c1-11e7-9ced-353e8ace8093.png)

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -523,51 +552,75 @@ AP_GPS::update_instance(uint8_t instance)
-            primary_instance = 0;
+
+            // copy the primary instance to the blended instance
+            state[GPS_MAX_RECEIVERS] = state[primary_instance];
```
You aren't blending here so why is the state being copied to the blended state location? The are other code paths that set _output_is_blended to false, but don't ever update the state array. (Same with the _output_location_corrected for that matter).

priseborough replied:
It's done for defensive programming reasons. It means that there is always something usable in the blended instance if it is inadvertently read.

WickedShell replied:
That's actually my point, there are other cases where we set output_is_blended to false, but I don't see any sign that we did the same defensive programming in those cases. I think if we are going to take that approach then all the cases should behave the same.

WickedShell replied:
Ah, I see actually, you are setting the blend weight on the primary to 1 in that scenario and letting it pass into the calc function. I missed that before.

WickedShell commented on libraries/AP_GPS/AP_GPS.h:
```diff
@@ -457,6 +473,28 @@ class AP_GPS
+    float _hgt_offset_cm[GPS_MAX_RECEIVERS]; // Filtered height offset from GPS instance relative to blended solution in _output_state.location (cm)
+    Vector3f _blended_antenna_offset; // blended antenna offset
+    float _blend_weights[GPS_MAX_RECEIVERS]; // blend weight for each GPS. The blend weights must sum to 1.0 across all instances.
+    Location _output_location_corrected; // output location after application of the  filtered offset corrections _NE_pos_offset_m and _hgt_offset_cm
```
What is this used for? I can see we do some math with it, but we never read it anywhere that I can find after it's had calculations done on it.

priseborough replied:
_hgt_offset_cm is used internally to calculate corrected_location[i].alt
_blended_antenna_offset is published by get_antenna_offset()
_blend_weight is used internally in many places

_output_location_corrected is not being used and will be removed.

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -226,6 +242,19 @@ void AP_GPS::init(DataFlash_Class *dataflash, const AP_SerialManager& serial_man
     _last_instance_swap_ms = 0;
+
+    // Initialise class variables used to do GPS blending
+    memset(&_NE_pos_offset_m, 0, sizeof(_NE_pos_offset_m));
```
These are already init'ed to 0 I believe, and failing that should be init'ed in the constructor, which minimizes flash consumption. @tridge might have some more detailed comments on that one, I admit I'm reciting what I've been informed of here.

priseborough replied:
They are declared, but not initialised elsewhere. I am following the method of initialisation used previously for other class variables.

I can't comment on the flash issue.

rmackay9 commented:
I tested this quite a bit today in SITL using a couple of commits that add a SIM_GP2_GLITCH vector: https://github.com/rmackay9/rmackay9-ardupilot/commits/prise-gpsblending

Overall it worked well but I found a couple of issues:

1. master isn't resetting the EKF position to the GPS position after a glitch.  Issue raised here: https://github.com/ArduPilot/ardupilot/issues/5813
2. if the GPS_AUTO_SWITCH =2 and EKF_ALT_SOURCE = 2, if one GPS is disabled (i.e. param set GPS_TYPE 0) and a position reset occurs, the altitude altitude does not immediately jump to the remaining GPS's altitude.  Instead it slowly moves towards it and the vehicle increases/decreases it's throttle during this period.  This behaviour is quite different from when GPS_AUTO_SWITCH=1 in which case the altitude resets immediately to the remaning GPS and there's almost no throttle change (only 2% change for a short time)
3. changing GPS_AUTO_SWITCH before takeoff led to an inability to get past the pre-arm checks (i.e. reboot required).  I.e. from 2 to 1 led to "EKF2 IMU0/1 has stopped aiding" messages and inability to arm in Loiter.  From 1 to 2 led to an immediate disarm before the vehicle could get off the ground (EKF failsafe?).
4. I was unable to make the blending use anything but 0/100, 50/50 or 100/0 blend.  I.e. I could never manage to make it use say 30% of one GPS and 70% of the other.  I probably don't know what values to change, I tried updating the horizontal position accuracy numbers (gps1 as 1500, gps2 as 3000) but it still did a 50/50 split.

Some things that worked well with GPS_AUTO_SWITCH set to either 1 or 2:

- disabling either GPS led to the horizontal position estimate immediately switching to the working GPS with no sudden movements of the vehicle.
-  the blended position was 1/2 way between the two GPSs (which was nice).

After discussion with @WickedShell, @priseborough, we agree to add a check that the GPS_RATE_MS is never more than 200ms.  We will likely add a start-up check that it's a reasonable value and add an accessor that regularly checks it's <200.

priseborough commented:
Randy, addressing your points in order:

1. This has been fixed by https://github.com/ArduPilot/ardupilot/pull/5816
2. I will investigate this further
3.  I will add a reboot required attribute to the parameter
4. I have seen ratios other than 0, 50 or 100 during bench and flight testing, however the blend ratio calculation uses both speed and horizontal position accuracy data by default. I will contact you for discussion on how the speed accuracy values can also be modified when testing in SITL.
5. Agree

cglusky commented:
Watching this closely.  Wanted to ask if this will have potential consequences for those of us testing/using RTK GNSS as one of two GNSS in a dual setup?  I am assuming if we want to make sure we are using best absolute position - assuming RTK is working and becomes primary during flight - that we should leave the parameter GPS_AUTO_SWITCH = 1 to use the original non blended mode/code?  I was not clear on how much impact RTK may have on the weighting.  I know there have been other issues raised for tracking RTK status but not sure where they stand, or if they ave impact on this issue.  Thanks.

EShamaev commented:
Just wondering, why not add two GPS sources to observation vector and integrate them both inherently in EKF?

priseborough commented:
@cglusky If the RTK receiver reports speed and/or position accuracy, then the much smaller values reported will make the weighting go to close to 100% for the RTK receiver when operating in RTK mode. Check the GPA message logged for the RTK receiver and see if there are non-zero HAcc and/or SAcc values. if so, then the blending algorithm can be used.

priseborough commented:
@EShamaev That requires adding additional states to the GPS to track the offset between receivers and additional  switching logic complexity. It would also have to be added to both EKF2 and EKF3. This would blow the flash budget, increase implementation and maintenance cost and potentially add bugs to the EKF's for a feature that only benefits a small percentage of users.

priseborough commented:
Looks like we're out of flash space on Pixhawk (again)

WickedShell commented:
@priseborough If someone would merge: https://github.com/ArduPilot/ardupilot/pull/5810 you should be fine...

proficnc commented:
Out of space on fmuv3 or v2, an important distinction.

rmackay9 commented:
Just merged @WickedShell's fix to reduce flash consumption by chopping solo gimbal.  It's a straightforward fix and no Solo's require the -v2 firmware (i.e. the cube doesn't suffer from the 1MB flash limit).  We just need to make sure that people know to load the -v3 firmware when using solo.

proficnc commented:
Awesome!

rmackay9 commented:
I've re-tested this with the fixes from @priseborough and it looks good to me now.  So I raised 5 points above and the results are:

1. master is resetting the horizontal position estimate after a glitch now (#5813)
2. the altitude estimate is resetting immediately now when GPS mix changes and EKF_ALT_SOURCE = 2.
3. changing GPS_AUTO_SWITCH before takeoff produced these results:
- 1 -> 2 : no bad behaviour
- 2 -> 1 : EKF variances led to immediate disarm after arming (which is a non-dangerous result)
4. still couldn't get any blend besides 0/100, 50/50 or 100/0 but I'm not too concerned.
5. we can add the pre-arm check that the two GPS are close together as a follow-up change.

I've got one extra small patch that I'd like included in the PR (https://github.com/rmackay9/rmackay9-ardupilot/commit/1c96d664ba813bf42bc360b9eca402590d1542fc).

Kiwa21 commented:
Just found this thread and following with interest since I was asking myself the same questions for fusing 3 GPS with an EKF to get a more accurate position. 

Quick bonus question, would it be possible to **force** the GPS MAV arming/home position to a known position as a parameter and then calculate offset from this position to correct the reading from the live GPS data ? ( say we have an external very precise RTK position and we are always taking off from the exact same point every time) ? I guess it would be the same strategy using GPS offset as you are doing for the blended virtual GPS ?

priseborough commented:
@Kiwa21 The algorithm that is used for the blending has been designed to support more than 2 GPS units by increasing the value of the GPS_MAX_RECEIVERS #define, although other parts of the GPS library would need to be modified to accomodate the additional receiver(s).

Telling the vehicle where it is accurately at the start of the flight is pointless unless the onboard GPS receivers have an equivalent accuracy, because the positional error offset of the receivers is not constant during flight. If there is an RTK receiver on-board, then the exisiting code is already capable of using that to perform precision recovery.

OXINARF commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+        }
+    }
+    // calculate output
+    if (!weeks_inconsistent) {
```
I think this should be `if (weeks_inconsistent)` to be consistent with the comments

rmackay9 replied:
Nice fine!  The double negative was confusing and led to the error.  We've reversed the var to be "weeks_consistent" and fixed the logic bug in the latest PR.

OXINARF commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+        }
+    }
+    timing[GPS_MAX_RECEIVERS].last_fix_time_ms = (uint32_t)temp_time_1;
+    timing[GPS_MAX_RECEIVERS].last_message_time_ms = (uint32_t)temp_time_2;
```
If we are blending these here there's no need to be setting them to the newest value above.

OXINARF commented on libraries/AP_GPS/AP_GPS.h:
```diff
@@ -391,10 +406,11 @@ class AP_GPS
-    AP_HAL::UARTDriver *_port[GPS_MAX_INSTANCES];
+    // Note allowance for an additional instance to contain blended data
+    GPS_timing timing[GPS_MAX_RECEIVERS+1];
+    GPS_State state[GPS_MAX_RECEIVERS+1];
```
Any reason not to use GPS_MAX_INSTANCES?

rmackay9 replied:
I think maybe it's clearer to have the +1 there?  Maybe not, I don't have a strong preference either way.

OXINARF replied:
I find the +1 more clearer too, but then I think we should just scrape the GPS_MAX_INSTANCES entirely (or even better, use it instead of creating a new GPS_MAX_RECEIVERS). The vehicles should really use the *num_sensors* method instead of relying on using the macro anyway.

OXINARF commented on libraries/AP_GPS/AP_GPS.h:
```diff
@@ -457,6 +473,27 @@ class AP_GPS
+// defines used to specify the mask position for use of different accuracy metrics in the blending algorithm
+#define USE_HPOS_ACC    1
+#define USE_VPOS_ACC    2
+#define USE_SPD_ACC     4
```
Nitpicking: these should be in cpp file to reduce scope.

rmackay9 replied:
ok, moving them.

OXINARF commented on libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp:
```diff
@@ -234,14 +236,6 @@ void NavEKF2_core::SelectVelPosFusion()
-        // Don't fuse velocity data if GPS doesn't support it
-        if (frontend->_fusionModeGPS <= 1) {
-            fuseVelData = true;
-        } else {
```
Why was this moved? Don't we want to fuse (or stop fusing) velocity data immediately?

rmackay9 replied:
I'll leave this to @priseborough to answer.

priseborough replied:
We still want to calculate a offset corrected GPS velocity measurement for other purposes, eg in-flight state resets, even if we aren't fusing the data, hence the changed order.

OXINARF commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    if (_blend_mask & USE_HPOS_ACC) {
+        for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+            if (state[i].status >= GPS_OK_FIX_2D) {
+                if (state[i].have_horizontal_accuracy && state[i].horizontal_accuracy > 0.0f) {
```
Check for bigger than 0 was removed from speed but not from vertical and horizontal accuracy.

rmackay9 replied:
@priseborough wants to check that the accuracy numbers are non-zero so I'm adding the check back in for speed-accuracy.  As far as I can see there's no where else that we specifically check that they are non-zero.

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -65,7 +65,8 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
+    // @Values: 0:Disabled,1:Hard,2:Soft
+    // @RebootRequired: True
     // @User: Advanced
     AP_GROUPINFO("AUTO_SWITCH", 3, AP_GPS, _auto_switch, 1),
```
@OXINARF unfortunately thats a bit orthogonal to the concept of bringing in the behavior disabled until confidence is gained on it across the board. Unless we are that confident in this?

OXINARF replied:
True, but after this is ready for master, I'm more confident in this method than the previous one but maybe that's just me. I'm fine with putting it to disabled by default too - it's just leaving it at the old method by default doesn't look good to me.

rmackay9 replied:
I think Tridge and others will be happier if we leave the existing "UseBest" logic (that's what I'm calling it) for a while until the "Blending" has had more testing by use and beta testers.

OXINARF replied:
Ok, if that's what the majority prefers. My issue is that UseBest doesn't always uses the best one (and that's one of the reasons we've recommended against using it).

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    // zero the blend weights
+    memset(&_blend_weights, 0, sizeof(_blend_weights));
+
+    // determine how many GPS receivers are returning a fix and the instance of the first receiver that is
```
The first receiver that is returning a fix I assume.

rmackay9 replied:
I've fixed this comment

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+        if (state[i].status > NO_FIX) {
+            num_instances = i+1;
+            if (first_instance == 0) {
+                first_instance = i;
```
Having the code written that way will scale to blending 2 receivers if we enable 3 GPS receivers, I think its a worthwhile check.

OXINARF replied:
I'm not following you. Unless I missed something *first_instance* is only used when you have less than two GPS receivers.

rmackay9 replied:
OK, I've removed the first_instance loop because the only case where it can be used is the case where it's value is zero.

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    uint8_t first_instance = 0;
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+        if (state[i].status > NO_FIX) {
+            num_instances = i+1;
```
Okay thats definetly true, num_instances is definitely not something blending should be messing with, it's fine for blending to have it's own version, but altering the value of num_instances is really bad, and that shouldn't be happening here.

rmackay9 replied:
OK, i'm moving it to near the top of the update function.
By the way, I will create a new PR rebased on master shortly.

WickedShell replied:
@rmackay9 It's already managed, you can just delete this section that change num_instances. Or if its critical (I think the blendings interpretation is different, which means it should be a different variable) rename this function.

WickedShell commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    // calculate the sum squared speed accuracy across all GPS sensors
+    float speed_accuracy_sum_sq = 0.0f;
+    if (_blend_mask & USE_SPD_ACC) {
+        for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
```
Not necessarily, if you considered 4 GPS instances, with only 2 instances its possible instance 1 and 3 are the ones that have actual ones. I think this is the more future proof solution, and at the moment will be an equal run time to num_instances.

OXINARF replied:
If you have instances 1 and 3 then *num_instances* will be 4 (if are counting from 1)...

WickedShell replied:
Doh! I forgot about that... I always think of num_instances as counting the number of loaded drivers... We'll probably need to update that later when we support more GPS instances (thats outside this scope though). I'm inclined to say that the GPS_MAX_RECEIVERS is still the correct implementation here though.

OXINARF replied:
I agree that changing *num_instances* to really be the number of available instances would be great, but there are several places that would need to be changed - and certainly when doing this we need a way to loop only through existing instances.

In this specific case, let's say we don't do that change but increase max receivers to 4 and a user only has the 2 first GPS instances, then we would have several _for_ loops looping 4 times instead of just 2, for no reason. I understand that you think that using GPS_MAX_RECEIVERS is better considering the future, but I think the future will need to have something better replacing *num_instances* that would be then replaced here.

Anyway, this isn't a bug and for now it's completely irrelevant as GPS_MAX_RECEIVERS is 2, so I'm not going to push it anymore :grinning:

WickedShell replied:
If num_instances changes to 2 then my scenario becomes active again :)

I agree this might need more thought later, however I think we can agree that inefficiency aside this will be valid for the future safely, while if we change the num_instances that might not be the case anymore. And as you stated atm it would be interchangable.

OXINARF replied:
> If num_instances changes to 2 then my scenario becomes active again :)

We already cleared that in your scenario *num_instances* would be 4 (if counting from 1). This is pretty clear, using *num_instances* (with it's current definition) would go through all existing instances and avoid any non-existing instance.

> however I think we can agree that inefficiency aside this will be valid for the future safely, while if we change the num_instances that might not be the case anymore.

I'd be worried if we changed *num_instances* in the future without checking all use-cases - it was actually being changed in this PR and I caught it :smiley:

rmackay9 replied:
I'm going to leave this using GPS_MAX_RECEIVERS...

OXINARF commented:
I forgot one thing in my review: all methods that receive an instance number need to be re-checked. At least *get_lag* looks like it won't work as it is accessing wrong memory positions.

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -65,7 +65,8 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
     // @DisplayName: Automatic Switchover Setting
     // @Description: Automatic switchover to GPS reporting best lock
-    // @Values: 0:Disabled,1:Enabled
+    // @Values: 0:Disabled,1:Hard,2:Soft
```
OK, i'm fixing up the AUTO_SWITCH parameter description

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -213,6 +214,22 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
+
+    // @Param: BLEND_MASK
+    // @DisplayName: Multi GPS Blending Mask
+    // @Description: Determines which of the accuracy measures Horizontal position, Vertical Position and Speed are used to calculate the weighting on each GPS receiver when soft switching has been selected by setting GPS_AUTO_SWITCH to 3.
```
I've fix the BLEND_MASK parameter description as well in my latest commits.

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -523,51 +552,74 @@ AP_GPS::update_instance(uint8_t instance)
+
+            // copy the primary instance to the blended instance
+            state[GPS_MAX_RECEIVERS] = state[primary_instance];
+            _blended_antenna_offset = _antenna_offset[primary_instance];
```
OK, i'm moving this outside the loop... or more accurately I will move the for loop within the "if (_auto_switch==1)" check.

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    }
+
+    // If not enough receivers to do blending or blending has  been disabled, set the primary instance and weighting to the first receiver detected
+    if (num_instances < 2 || _auto_switch != 2) {
```
I've removed this redundant check of _auto_switch.

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+        // receiver data has timed out so hard switch to the newest data
+        state[GPS_MAX_RECEIVERS].last_gps_time_ms = max_ms;
+        // hard switch to that receiver
+        primary_instance = newest_instance;
```
Perhaps @priseborough can answer better but I suspect the point is that we've missed a GPS message from one of the GPSs.  Our choice is either to simply not provide an update at all (to the EKF, etc) or to use the one that's providing data.  So it's not really a 2D vs 3D question.. it's a "has data" vs "does not have data" question.

OXINARF replied:
@rmackay9 I understand how it is choosing, I'm saying that doesn't look the best method to me. A 2D position isn't used by anyone (as far as I'm aware), but an old 3D position can still be used.
Again, this isn't so important now because we only have two GPS instances but when that number is increased, it can happen that it chooses a GPS with 2D fix when it has another with 3D fix that just received its last message one millisecond earlier (and so wasn't the chosen one).

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    // Check if we can do blending using reported accuracy
+    bool can_do_blending = (horizontal_accuracy_sum_sq > 0.0f || vertical_accuracy_sum_sq > 0.0f || speed_accuracy_sum_sq > 0.0f);
+
+    // if we cant do blending using reported accuracy, set the weight to the receiver with the highest solution status
```
I've attempted to make it fall back in my latest set of commits.

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    // A weighting of 1 will make the offset adjust the slowest, a weighting of 0 will make it adjust with zero filtering
+    float alpha[GPS_MAX_RECEIVERS];
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+        if (state[i].last_gps_time_ms > _last_time_updated[i]) {
```
yes, I think you're right.  Darn, I'm going to have to leave @priseborough to resolve that.

rmackay9 replied:
_last_time_updated issue fixed in the updated PR.

rmackay9 commented:
Well, I've done something which addresses many of the issues found by @OXINARF but not quite all of them.  The commit history is a bit messy though so perhaps @priseborough and I should do another review tomorrow before we ask anyone to look at it again.
https://github.com/rmackay9/rmackay9-ardupilot/commits/prise-gpsblending2

I've lightly tested it and it doesn't seem broken.

magicrub commented:
related to https://github.com/ArduPilot/ardupilot/issues/3754 and and https://github.com/ArduPilot/ardupilot/issues/1597

OXINARF commented:
@rmackay9 It would be great if history could be a bit squashed, it's easier to check individual commits, but not when there are several fixup commits.

rmackay9 commented on libraries/AP_GPS/AP_GPS.h:
```diff
@@ -28,7 +28,8 @@
    than 1 then redundant sensors may be available
  */
-#define GPS_MAX_INSTANCES 2
+#define GPS_MAX_RECEIVERS 2 // maximum number of physical GPS sensors allowed - does not additional 'virtual' GPS created by blending receiver data
```
fixed in the updated PR

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+
+    // Calculate filter coefficients to be applied to the offsets for each GPS position and height offset
+    // A weighting of 1 will make the offset adjust the slowest, a weighting of 0 will make it adjust with zero filtering
+    float alpha[GPS_MAX_RECEIVERS];
```
alpha's initialisation is fixed in the updated PR

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+        }
+    } else {
+        for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+            hpos_blend_weights[i] = 0.0f;
```
this initialisation is made more efficient in the updated PR.

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
+    blended_NE_offset_m.zero();
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+        if (_blend_weights[i] > 0.0f && i != best_index) {
+            //TODO correct for antenna position
```
Yes, TODO removed in the latest PR.

rmackay9 commented:
@priseborough and I did some more work on this today and have created a new PR: https://github.com/ArduPilot/ardupilot/pull/5841

I'll close this one for now.  Thanks for the reviews!
